### PR TITLE
Initial stylesheet styleguide

### PR DESCRIPTION
**Pull request by dextermb**
## Description
As requested by @wheatleyjj I have put together a general overview on how I think stylesheets should be structured and written. Hopefully everyone agrees. If there are any queries then comment in this thread so we can discuss them. 👍 🇬🇧 

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds documentation)
- [ ] Breaking change (This is a significant change)

## Checklist

- [x] I have ensured that I have used relative links where appropriate.
- [x] I have proofread my documentation.

### CODE REVIEW

File: styleguides/stylesheets/core.md [hunk 1]
```diff
@@ -0,0 +1,355 @@
...
+    text-transform:   capitalize;
+}
+```
+
+##### Multi-value properties should go below the property
+
+Bad:
+
+```css
+.button.button_danger {
```
Comment (DivineOmega) [inline styleguides/stylesheets/core.md]: This should be kebab case here as you're not demonstrating the case difference here.

```suggestion
.button.button-danger {
```

File: styleguides/stylesheets/core.md [hunk 2]
```diff
@@ -0,0 +1,355 @@
...
+Good:
+
+```css
+.button.button-danger {
+    background-color: red;
+    text-transform:   capitalize;
+}
+```
+
+##### Colors should be readable
```
Comment (DivineOmega) [inline styleguides/stylesheets/core.md]: What is considered readable is not clear. Perhaps reword this to be more obvious?

File: styleguides/stylesheets/core.md [hunk 3]
```diff
@@ -0,0 +1,355 @@
...
+
+```css
+.button.button-danger {
+    background-color: red;
+    text-shadow:      0 2px 4px #ff0;
+    color:            rgba(0, 0, 0, 1); // rgb(0, 0, 0)
+}
+```
+
+##### [Conbinators][css-conbinators] should have spaces around them
```
Comment (DivineOmega) [inline styleguides/stylesheets/core.md]:
```suggestion
##### [Combinators][css-combinators] should have spaces around them
```

Spelling.

File: styleguides/stylesheets/core.md [hunk 4]
```diff
@@ -0,0 +1,355 @@
...
+    .button {
+        background: $button-background-color;
+    }
+```
+
+<small><i>Note: The variables file must be imported before anything otherwise a parse error will happen.</i></small>
+
+
+[kebab-case]: https://stackoverflow.com/a/12273101/8176262
+[css-conbinators]: https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Selectors#Combinators
```
Comment (DivineOmega) [inline styleguides/stylesheets/core.md]:
```suggestion
[css-combinators]: https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Selectors#Combinators
```

Spelling.

File: styleguides/stylesheets/core.md (hunk 2 above)
Comment (dextermb) [reply styleguides/stylesheets/core.md]: Ah! It seems like the good example got reset. Let me re-commit this.

Comment (dextermb) [reply styleguides/stylesheets/core.md]: I define readable by having full length hexcodes, and spacing between `rgb`/`rgba` color parameters.

Comment (DivineOmega) [reply styleguides/stylesheets/core.md]: Makes sense. It might be worth bullet pointing these.

File: styleguides/stylesheets/core.md [hunk 5]
```diff
@@ -0,0 +1,360 @@
...
+.button.button-danger {
+    background-color: red;
+    text-transform:   capitalize;
+}
+```
+
+##### Colors should be readable
+
+- Named colors should be lowercase
+- Hex codes should be full length
```
Comment (DivineOmega) [inline styleguides/stylesheets/core.md]:
```suggestion
- Hex colours should be full length.
```

File: styleguides/stylesheets/core.md [hunk 6]
```diff
@@ -0,0 +1,360 @@
...
+- Hex codes should be uppercase
```
Comment (DivineOmega) [inline styleguides/stylesheets/core.md]:
```suggestion
- Hex colours should be uppercase.
```

File: styleguides/stylesheets/core.md [hunk 7]
```diff
@@ -0,0 +1,360 @@
...
+```css
+.button.button-danger {
+    background-color: red;
+    text-transform:   capitalize;
+}
+```
+
+##### Colors should be readable
+
+- Named colors should be lowercase
```
Comment (DivineOmega) [inline styleguides/stylesheets/core.md]:
```suggestion
- Named colours should be lowercase.
```

File: styleguides/stylesheets/core.md [hunk 8]
```diff
@@ -0,0 +1,360 @@
...
+- Hex codes should be full length
+- Hex codes should be uppercase
+- RGB/RGBA colors should have spaces between parameters
```
Comment (DivineOmega) [inline styleguides/stylesheets/core.md]:
```suggestion
- RGB/RGBA colours should have spaces between parameters.
```

File: styleguides/stylesheets/core.md [hunk 9]
```diff
@@ -0,0 +1,360 @@
...
+Good:
+
+```css
+.button.button-danger {
+    background-color: red;
+    text-transform:   capitalize;
+}
+```
+
+##### Colors should be readable
```
Comment (DivineOmega) [inline styleguides/stylesheets/core.md]:
```suggestion
##### Colours should be readable
```

**Comment by AlexCatch** [comment]
Should standard methodologies be included for the style guide? Like one of the below?

https://www.webfx.com/blog/web-design/css-methodologies/

File: styleguides/stylesheets/core.md [hunk 10]
```diff
@@ -0,0 +1,360 @@
...
+.button.danger {
+    background-color: red;
+    text-transform:   capitalize;
+}
+```
+
+Good:
+
+```css
+.button.button-danger {
```
[the remark this answers is not in the document]
Comment (DivineOmega) [reply styleguides/stylesheets/core.md]: Regardless of which methodology we go with, I believe this point should be clarified in the style guide ideally.

File: styleguides/stylesheets/core.md [hunk 11]
```diff
@@ -0,0 +1,360 @@
...
+```
+
+<small><i>Note: [`flex`][flex] can cover most situations where `calc` would be used.</i></small>
+
+
+## LESS/SCSS
+
+In these examples variables will be used as shown in [SCSS][scss-variables]. If you are using [LESS][less-variables] make sure to change the syntax accordingly.
+
+##### Styles should be broken up into entities and imported into a single file
```
[the remark this answers is not in the document]
Comment (dextermb) [reply styleguides/stylesheets/core.md]: For sure. But currently we don't have anything like that. Once we sit down and think about file structures for _everything else_ then we can pull this section out and move it into there.

File: styleguides/stylesheets/core.md [hunk 12]
```diff
@@ -0,0 +1,360 @@
...
+
+// ...
+
+.button {
+	background-color: $button-background-color;
+}
+
+```
+
+##### Variables should go into their own file
```
[the remark this answers is not in the document]
Comment (dextermb) [reply styleguides/stylesheets/core.md]: For sure. But currently we don't have anything like that. Once we sit down and think about file structures for _everything else_ then we can pull this section out and move it into there.

**Comment by AlexCatch** [comment]
While this is a great step forward and we do need a CSS style guide, I think it's important to seperate architectural decisions and style decisions from each other, style guides should specify how you format your code, it shouldn't specify how a project is structured or where code should go.

It might be worth splitting a few sections into a `CSS best practices` file.

File: styleguides/stylesheets/core.md [hunk 13]
```diff
@@ -0,0 +1,360 @@
...
+    .button {}
+
+- styles/containers.scss
+
+    body {}
+    .page {}
+
+```
+
+##### Variables should be prefixed and have the property within the name
```
[the remark this answers is not in the document]
Comment (dextermb) [reply styleguides/stylesheets/core.md]: How I see the example working would be:

```scss
$red: red;

$button-color: $red;
$header-color: $red;
```

This way if you wanted to change the _red_ to a darker or lighter color then you would only have to edit it in once place.

File: styleguides/stylesheets/core.md (hunk 10 above)
[the remark this answers is not in the document]
Comment (dextermb) [reply styleguides/stylesheets/core.md]: Similarly to #87. It's best to get something in and then add finer details at a later stage.

[the remark this answers is not in the document]
Comment (AlexCatch) [reply styleguides/stylesheets/core.md]: It does look like it's typically used in all methodologies so i'm happy.

File: styleguides/stylesheets/core.md [hunk 14]
```diff
@@ -0,0 +1,360 @@
...
+button {
+    background-color: red;
+    text-transform:   capitalize;
+}
+```
+
+<small><i>Note: A tab with the size of 4 spaces will also work.</i></small>
+
+
+##### Property values should be aligned by value
```
[the remark this answers is not in the document]
Comment (dextermb) [reply styleguides/stylesheets/core.md]: Certainly up for discussion- I'll create an issue on this for you.

**Comment by DivineOmega** [comment]
@AlexCatch As you've suggested before, I think it might be best to get something in place quickly and then we can look at restructuring it at a later point. What do you think?

File: styleguides/stylesheets/core.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (AlexCatch) [reply styleguides/stylesheets/core.md]: But what is the difference between that and 

```css
$red: red;

.header {
	color: $red;
}
```

It just seems like an unnecessary variable to me unless I'm looking at it wrong

```css
$red: red;
$headerColor: red;

.header {
	color: $headerColor;
}
```

**Comment by AlexCatch** [comment]
@DivineOmega I'm happy with it as long as we're in agreement that they should ideally be separated in the future.

File: styleguides/stylesheets/core.md (hunk 14 above)
[the remark this answers is not in the document]
Comment (dextermb) [reply styleguides/stylesheets/core.md]: Here's the issue: #92 🇬🇧

File: styleguides/stylesheets/core.md (hunk 13 above)
[the remark this answers is not in the document]
Comment (dextermb) [reply styleguides/stylesheets/core.md]: I think you'd only create _color variable_ such as `$red` if you were looking to use a color in multiple places. For example: `$brand-color: red;`.

**Comment by DivineOmega** [comment]
@AlexCatch It definitely makes sense to put the guidelines in the most appropriate location.

**Comment by dextermb** [comment]
> @DivineOmega I'm happy with it as long as we're in agreement that they should ideally be separated in the future.

I agree especially as if you are planning to do some other file structure related PRs in the future.
